stubs/python-jose/jose/backends/base.pyi

```diff
@@ -1,17 +1,23 @@
+from _typeshed import Self
 from typing import Any
 
 class Key:
+    # Enable when we can use stubs from installed dependencies,
+    # as `key` can be of type cryptography.x509.base.Certificate:
+    # from cryptography.x509 import Certificate
     def __init__(self, key, algorithm) -> None: ...
-    def sign(self, msg) -> None: ...
-    def verify(self, msg, sig) -> None: ...
-    def public_key(self) -> None: ...
-    def to_pem(self) -> None: ...
-    def to_dict(self) -> None: ...
-    def encrypt(self, plain_text, aad: Any | None = ...) -> None: ...
-    def decrypt(self, cipher_text, iv: Any | None = ..., aad: Any | None = ..., tag: Any | None = ...) -> None: ...
-    def wrap_key(self, key_data) -> None: ...
-    def unwrap_key(self, wrapped_key) -> None: ...
+    def sign(self, msg: bytes) -> bytes: ...
+    def verify(self, msg: bytes, sig: bytes) -> bool: ...
+    def public_key(self: Self) -> Self: ...
+    def to_pem(self) -> bytes: ...
+    def to_dict(self) -> dict[str, Any]: ...
+    def encrypt(self, plain_text: str | bytes, aad: bytes | None = ...) -> tuple[bytes, bytes, bytes | None]: ...
+    def decrypt(
+        self, cipher_text: str | bytes, iv: str | bytes | None = ..., aad: bytes | None = ..., tag: bytes = ...
```

Comment:
How do we usually go about cases where the function has a `None` default, as it happens with `tag` here, but then it checks it must not be `None`? https://github.com/mpdavis/python-jose/blob/master/jose/backends/cryptography_backend.py#L496-L497
Tests are currently failing due to this issue https://github.com/python/typeshed/runs/7339240425?check_suite_focus=true

Reply:
I think stubtest is raising a fair point here. It looks to me like `ValueError` is only raised if `tag` is `None` _and_ `mode.name == "GCM"`. Presumably if `mode.name` is equal to something else, passing in `None` for the `tag` parameter is fine. https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/backends/cryptography_backend.py#L495

Reply:
While decrypting with AES-GCM makes sense without a tag, it's usually either a mistake or a part of some quick-and-dirty script, as there's no guarantee that you decrypted the data you intended to decrypt.

But maybe we should disallow passing a tag altogether when mode isn't GCM. Googling "aes tag", every result mentions GCM which makes me believe that tags are purely an AES-GCM thing.

Reply:
And now that I look at the code, it doesn't use `tag` at all when mode isn't GCM. So I think there should be two overloads:
- Mode GCM, tag specified as not None (this actually needs to be split into two overloads: tag passed positionally, tag passed by keyword)
- Any other mode, tag None or unspecified

But this too can be improved in a separate PR, if it needs to be done at all :)

Reply:
I don't think we can do much typing based on IV values: it's just a `bytes` object (possibly `str`?), of length unknown at type checking time. Either way, this is beyond the scope of this PR and we should probably just merge this as is :)